tests/unit/test_pipelines.cpp

```diff
@@ -0,0 +1,262 @@
+
+
+using namespace mosaic::pipeline;
+
+class IntProducer: public Box, public Producer<int> {
+public:
+  IntProducer(int i, std::function<void()> pipeline_stop_callback) :
+    m_i(i),
+    m_pipeline_stop_callback(pipeline_stop_callback)
+  {}
+
+  virtual ~IntProducer() {
+    stop();
+  }
+
+  void pre_start() override {
+    m_pre_start_called = true;
+  }
+
+  void start() override {
+    m_start_called = true;
+    m_thread = std::thread([&]() { thread_body(); });
+  }
+
+  void stop() override {
+    m_stop_called = true;
+    if (m_thread.joinable())
+      m_thread.join();
+  }
+
+  bool m_pre_start_called{false};
+  bool m_start_called{false};
+  bool m_stop_called{false};
+
+private:
+  void thread_body() {
+    for (int i = m_i; i <= m_i + 5; ++i) {
+      // Waiting for a while will guarantee the predictable order on the receiving side (consumer)
+      // and it can be therefore tested.
+      std::this_thread::sleep_for(std::chrono::milliseconds{50});
+      produce(i);
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds{50});
+    m_pipeline_stop_callback();
+  }
+
+  int m_i;
+  std::function<void()> m_pipeline_stop_callback;
+  std::thread m_thread;
+};
+
+class IntConsumer: public Box, public Consumer<int> {
+
+
+  void pre_start() override {
+    m_pre_start_called = true;
+  }
+
+  void start() override {
+    m_start_called = true;
+  }
+
+  void stop() override {
+    m_stop_called = true;
+  }
+
+  void input(int v) override {
+    m_consumed_values.push_back(v);
+  }
+
+public:
+  virtual ~IntConsumer() = default;
+
+  std::vector<int> m_consumed_values;
+  bool m_start_called{false};
+  bool m_pre_start_called{false};
+  bool m_stop_called{false};
+};
+
+class IntDoubler: public Box, public Consumer<int>, public Producer<int> {
+
+  void pre_start() override {
+    m_pre_start_called = true;
+  }
+
+  void start() override {
+    m_start_called = true;
+  }
+
+  void stop() override {
+    m_stop_called = true;
+  }
+
+  void input(int v) override {
+      produce(2 * v);
+  }
+
+public:
+  virtual ~IntDoubler() = default;
+
+  bool m_start_called{false};
+  bool m_pre_start_called{false};
+  bool m_stop_called{false};
+};
+
+class VectorProducer: public Box, public Producer<std::vector<int>> {
+public:
+  VectorProducer(int i, std::function<void()> pipeline_stop_callback) :
+    m_i(i),
+    m_pipeline_stop_callback(pipeline_stop_callback)
+  {}
+
+  virtual ~VectorProducer()
+  {
+    stop();
+  }
+
+  void start() override {
+    m_thread = std::thread([&]() { thread_body(); });
+  }
+
+  void stop() override {
+    if (m_thread.joinable())
+      m_thread.join();
+  }
+
+private:
+  void thread_body() {
+    for (auto i = 0; i < m_i; i++)
+    {
+      produce({i, i+1, i+2});
+    }
+    m_pipeline_stop_callback();
+  }
+
+  int m_i;
+  std::function<void()> m_pipeline_stop_callback;
+  std::thread m_thread;
+};
+
+class VectorConsumer: public Box, public Consumer<std::vector<int>> {
```

Comment:
Tohle sice bude fungovat, ale puvodni myslenka byla ze zpravy jsou levne na kopirovani -- neco jako shared_ptr. Aktualne vse predavame hodnotou. Co si o tom myslis? Chceme to nejak vynucovat? Upravit API aby se nedelaly kopie?

Reply:
Tak mozna v tomto testu a obecne takovych pripadech pouzit `Consumer<std::shared_ptr<std::vector<int>>>`, ale nechat ten typ jako obecnou hodnotu `T?`

Reply:
S tim ze by si uzivatel sam rozhodl, jestli chce posilat jen jednoduchy typ nebo shared_ptr nejakeho bufferu.

Reply:
Jasne, tak to funguje ted, uzivatel se musi rozhodnout. Ale jak se ukazuje tady tak neni obvious ze velke hodnoty nebudou fungovat dobre. Ale teda nevim jak to elegantne zdokumentovat  v tom interface

Reply:
(v Rustu bychom vyzadovaly aby typ byl `Copy`)

Reply:
K tomu jsou v C++ "analogicke" concepts, v tomto pripade https://en.cppreference.com/w/cpp/concepts/copyable
,ale je to az od C++20.

Reply:
Muzu to zatim zduraznit v doxygen komentari.